---
title: "ATACseq_Session_1"
date: "2023-01-23"
output:
  html_document:
    df_print: paged
---

---
</br>

### **ATAC Session 1**
</br>

In this session we will begin analysing bulk ATACseq analysis in R. We will import mapped reads in the BAM format and create some QC plots to check our data. 

</br>

#### **Setup working environment**

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width=8, fig.height=8) 
```

In order to run our analysis in R, we first need to install some R packages. These packages include functions that we will use to manipulate and visualise our ATAC-seq data.

Required packages:
- 'ATACseqQC' - used to generate quality control plots
- 'ggplot2' - used to generate plots
- 'soGGi' - used to generate quality control plots

Required helper packages:
- 'BiocManager' - a package that helps us install the above packages

```{r, message=FALSE, warning=FALSE}
if (!require("tidyverse", quietly = TRUE))
    install.packages("tidyverse", repos = "http://cran.us.r-project.org")

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager", repos = "http://cran.us.r-project.org")

BiocManager::install(version = "3.16")
BiocManager::install("ATACseqQC")
BiocManager::install("soGGi")

install.packages('ggplot2', repos='http://cran.us.r-project.org')
```

Load packages

```{r, message=FALSE, warning=FALSE}
library(ATACseqQC)
library(soGGi)
library(ggplot2)
```

</br>

#### **Input data**

The data we will be using is from this paper:

[An Improved ATAC-seq Protocol Reduces Background and Enables Interrogation of Frozen Tissues. M Ryan Corces et al.](https://www.nature.com/articles/nmeth.4396)

This paper presents a new ATAC-seq protocol which reduces background and enables interrogation of frozen tissues. The human cell line GM12878 was cultured and ATACseq data was collected using 3 different protocols: Standard, Omni and Fast. For each protocol two replicates were collected. The ATAC-seq was performed using paired-end sequencing. 

The ATAC-seq reads were aligned and annotated for the hg19 reference genome using the [NF-core ATACseq pipeline](https://nf-co.re/atacseq). 

Here is the sample information:

```{r make_table, include=FALSE}
data <- data.frame(SRA_ID = c('SRR5427884', 'SRR5427885', 'SRR5427886', 'SRR5427887', 'SRR5427888', 'SRR5427889'), 
                   Method = c('Standard', 'Standard', 'Omni', 'Omni', 'Fast', 'Fast'),
                   Replicate = c(1, 2, 1, 2, 1, 2))
```

```{r print_table}
data
```

#### **Importing aligned reads**

First we need to import our BAM files and BAM index files. A BAM file contains information on aligned sequences. A BAM index file acts like a table of contents to make it easier to navigate a corresponding BAM file. 

As these data files are very big we are just going to look at a subset of the original data. Let's read in 2 files: one replicate of ATAC-seq data made using the 'OMNI' method and one replicate of data made using the 'STANDARD' method. 

First we need to read in this data, to do this we need to set the paths which tell us where this data is located. 
You will need to customise the following paths to reflect where you have save the data on your computer:

```{r set_paths}

# set working directory
setwd("~/dev/NGS_innovation_scholars")

# setting the paths for the OMNI data
OMNI_bam_path <- "./raw_data/BAM_files/OMNI.bam"
OMNI_bai_path <- "./raw_data/BAM_files/OMNI.bam.bai"

# setting the paths for the STD data
STD_bam_path <- "./raw_data/BAM_files/STD.bam"
STD_bai_path <- "./raw_data/BAM_files/STD.bam.bai"

```


### Check BAMs

We can use functions from the package ['ATACseqQC'](https://haibol2016.github.io/ATACseqQCWorkshop/articles/ATACseqQC_workshop.html) to inspect our BAM files and generate some ATAC-specific QC plots.

First, lets check our BAMs using the function `bamQC`:

```{r check_BAMs_OMNI}

bamQC(OMNI_bam_path, index = OMNI_bai_path, outPath = NULL)

```

```{r check_BAMs_STD}

bamQC(STD_bam_path, index = STD_bai_path, outPath = NULL)

```

##### Q: The BAM files have been subsetted to make them smaller and easier to work with, from the above output can you tell how they have been cut down? 
<details>
  <summary>Click for Answer</summary>
    A: We can see that, for both BAM files, there are only reads mapped to chromosome Y. The BAM files have been subsetted to only include this chromosome. 
</details> 


#### **Shifting reads**

A 9 base pair (bp) DNA insertion is introduced when the transposase-induced nick is repaired during the ATAC process. For some types of analysis we may want to have single base pair resolution. This means we will need to correct for this insertion by shifting our read coordinates. Reads aligned to the +ve strand should be offset by +4bp and to the -ve strand should be shifted by -5bp. For our purposes this shifting will not be necessary, but if needed this can done by using the ATACseqQC function `shiftGAlignmentsList` 


#### **Quality Control**

We have already had a look at the quality of our ATAC-seq data in the multi_qc report. After reads have been preprocessed and aligned, there are some additional ATAC-specific quality control visualisations we can generate. Check out this [review](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-1929-3) for a clear explanation of ATAC-specific quality control steps. 


### Fragment Size Distribution

Good-quality ATACseq data should display a characteristic pattern when you plot the distribution of the fragment sizes. This is because DNA is packaged into nucleosomes, which are about 147bp in length and cannot be accessed by the Tn5 enzyme. This means that most Tn5 fragments  come from nucleosome-free regions (NFR), and therefore be <100bp long. Fragments can also be created if Tn5 enzymes insert just before and after 1, 2 or 3 nucleosomes. This results in an enrichment of fragments that are ~200, ~400 and ~600bp long. 

First we can look at this distribution for the OMNI sample:

```{r frag_size_OMNI}

fragSize_OMNI <- fragSizeDist(OMNI_bam_path, "OMNI", index = OMNI_bai_path)

```


Then in our STD sample:

```{r frag_size_STD}

fragSize_STD <- fragSizeDist(STD_bam_path, "STD", index = STD_bai_path)

```

As we are only looking at a subset of the data, these plots are not perfect, but we can see a promising hint of the typical nucleosome banding pattern seen in good quality ATAC-seq data. 

If you take a look at the [MultiQC report](https://moodle.learninghub.kingshealthpartners.org/pluginfile.php/184577/mod_resource/content/3/multiqc_report.html#picard-insertsize) for the ATAC-seq data in the section 'Read Alignment and Quality Control' you can see that there is also a fragment distribution plot here for each of the samples. You can see on this plot that the OMNI samples have a much clearer fragment length distribution pattern than the STD samples. 

##### Q: The fragment distribution plots on the MultiQC report for the STD and OMNI samples look a little different from ours, why might that be?
<details>
  <summary>Click for Answer</summary>
    A: One explanation could be that we are only using the reads that map to chrosome Y to save memory, whilst the MultiQC report plot is likely to use reads from all chromosomes. 
</details> 

### Loading annotations

Annotations are needed so we can see where the genes are in the genome. They can help us identify if our fragments map to gene bodies, transcription start sites, introns or intergenic regions. As we are working with human data, we can download the annotations from the online database [UCSC](https://genome.ucsc.edu/).

```{r load_TSS_annotations, results = 'hide', message=FALSE, warning=FALSE}

# Load the human genome TxDB object
BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene")
library(TxDb.Hsapiens.UCSC.hg19.knownGene)

# Extract the gene locations
genesLocations <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)

# Extract the transcription start site (TSS) locations
tssLocations <- resize(genesLocations, fix="start", width = 1)

print(tssLocations)

```


###### Challenge: As our BAM files only include reads from the Y chromosome, we need to cut down our annotation file to also only include genes from this chromosome. Have a go at doing this yourself. 

<details>
  <summary>Click for Answer</summary>
  
`tssLocations` is a GRanges object, so in order to access the chromosome names from this object we can use the function `seqnames`.
  
```{r subset_annotations}
tssLocations <- tssLocations[seqnames(tssLocations) %in% "chrY"]
print(tssLocations)
```
</details> 

### Transcription Start Site Enrichment

Another sign of good-quality ATACseq data is that most reads should map to transcription start sites (TSSs). This is because TSSs are transcriptionally active and therefore 'open'. We can use the function `regionPlot` from the package 'soGGi' to examine this. 

``` {r TSS_enrichment_OMNI}

OMNI_plot <- regionPlot(bamFile = OMNI_bam_path, testRanges = tssLocations, paired = TRUE)

plotRegion(OMNI_plot) + theme_bw()

```

``` {r TSS_enrichment_STD}

STD_plot <- regionPlot(bamFile = STD_bam_path, testRanges = tssLocations, paired = TRUE)

plotRegion(OMNI_plot) + theme_bw()

```


If you take a look at another section of the [MultiQC report](https://moodle.learninghub.kingshealthpartners.org/pluginfile.php/184577/mod_resource/content/3/multiqc_report.html#mlib_deeptools_read_distribution_profile_plot) for the ATAC-seq data, you can see that there is also a TSS enrichment plot here. This plot is slightly different from ours in that, not only are all the TSSs aligned at the '0' point on the x axis, but their corresponding gene lengths have been adjusted so all the gene bodies end at the '1000' mark on the x axis. We can see that each sample is enriched at the TSSs. 




