---
title: "ATACSeq_Session_2"
date: "2023-03-09"
output:
  html_document:
    df_print: paged
---

</br>

### **ATAC Session 2**
</br>

We will conduct bulk ATACseq analysis in R using [DESeq2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8) for data normalisation and differential accessibility testing.

This tutorial is focussed on the data objects used during DESeq2 differential expression analysis.

</br>

#### **Setup working environment**

First we need to set up our working environment, including some parameters for the plots we will make later.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width=8, fig.height=8) 
```

We first need to install some R packages. These packages include functions that we will use to manipulate and visualise our ATAC-seq data.  

Required packages:  
- 'dplyr' and 'data.table' - packages for handling dataframes.    
- DESeq2 - package to run differential expression (or for ATAC accessibility) tests.   

Required helper packages:  
- 'BiocManager' - a package that helps us install the above packages. 

```{r installation, results = 'hide', message=FALSE, warning=FALSE}

# Set up params for installation
r <- getOption("repos")
r["CRAN"] <- "http://cran.cnr.berkeley.edu/"
options(repos = r)
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager", repos = "http://cran.us.r-project.org")

# Install BiocManager
BiocManager::install(version = "3.16")

# Install BiocManager packages
BiocManager::install("karyoploteR")
BiocManager::install("ELMER")
BiocManager::install("org.Hs.eg.db")

## Install CRAN packages


```

Load packages

```{r, message=FALSE, warning=FALSE}
library(data.table)
library(dplyr)
library(DESeq2)
library(ggplot2)
```

</br>

