---
title: "ATACseq_Session_1"
output: html_document
date: "2023-01-23"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Input data

The data we will be using is from this paper:
M Ryan Corces et al. An Improved ATAC-seq Protocol Reduces Background and Enables Interrogation of Frozen Tissues. Nat Methods. 2017 Oct;14(10):959-962. doi: 10.1038/nmeth.4396.

In this paper Corces et al present a new ATAC-seq protocol which reduces background and enables interrogation of frozen tissues. The human cell line GM12878 was cultured and ATACseq data was collected using 3 different protocols: Standard, Omni and Fast. For each protocol two replicates were collected. The ATAC-seq was performed using paired-end sequencing. 

Data was aligned and annotated for the hg19 reference genome using the NF-core ATACseq pipeline - https://nf-co.re/atacseq. 

Sample information:
SRA ID	    SAMPLE NAME

SRR5427884	ATAC-seq of in vitro culture GM12878 using the Standard ATAC method
SRR5427885	ATAC-seq of in vitro culture GM12878 using the Standard ATAC method

SRR5427886	ATAC-seq of in vitro culture GM12878 using the Omni-ATAC method
SRR5427887	ATAC-seq of in vitro culture GM12878 using the Omni-ATAC method

SRR5427888	ATAC-seq of in vitro culture GM12878 using the Fast-ATAC method
SRR5427889	ATAC-seq of in vitro culture GM12878 using the Fast-ATAC method

# Installing Packages

In order to run our analysis in R, we first need to install some R packages. These packages include functions that we will use to manipulate and visualise our ATAC-seq data.

Required packages:
- 'ATACseqQC' - used to generate quality control plots
- 'ggplot2' - used to generate plots
- 'soGGi' - used to generate quality control plots

Required helper packages:
- 'BiocManager' - a package that helps us install the above packages

```{r installation, echo = FALSE}

# Set up params for installation
r <- getOption("repos")
r["CRAN"] <- "http://cran.cnr.berkeley.edu/"
options(repos = r)

# Install BiocManager
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager", repos = "http://cran.us.r-project.org")
BiocManager::install(version = "3.16")

## Install ATACseqQC
BiocManager::install("ATACseqQC")

# Install ggplot2
install.packages("ggplot2")

# Install soGGi
BiocManager::install("soGGi")

```

# Importing aligned read data

First need to import our BAM files and BAM index files. BAM files contain information on aligned sequences. 

As these data files are very big we are just looking at a subset of the original data, only reads that map to the Y chromosome.

First we need to read in this data, to do this we need to set the paths which tell us where this data is located.  

```{r set_paths}

# setting the paths for the OMNI data
#OMNI_bam_path <- "https://raw.githubusercontent.com/alexthiery/NGS_training/main/data/NF-ATACseq_alignment/chrY_subset_bam/OMNI_chrY.bam"
#OMNI_bai_path <- "https://raw.githubusercontent.com/alexthiery/NGS_training/main/data/NF-ATACseq_alignment/chrY_subset_bam/OMNI_chrY.bam.bai"

# setting the paths for the OMNI data
OMNI_bam_path <- "./data/NF-ATACseq_alignment/chrY_subset_bam/OMNI_chrY.bam"
OMNI_bai_path <- "./data/NF-ATACseq_alignment/chrY_subset_bam/OMNI_chrY.bam.bai"

# setting the paths for the STD data
#STD_bam_path <- "https://raw.githubusercontent.com/alexthiery/NGS_training/main/data/NF-ATACseq_alignment/chrY_subset_bam/STD_chrY.bam"
#STD_bai_path <- "https://raw.githubusercontent.com/alexthiery/NGS_training/main/data/NF-ATACseq_alignment/chrY_subset_bam/STD_chrY.bam.bai"

# setting the paths for the STD data
STD_bam_path <- "./data/NF-ATACseq_alignment/chrY_subset_bam/STD_chrY.bam"
STD_bai_path <- "./data/NF-ATACseq_alignment/chrY_subset_bam/STD_chrY.bam.bai"


```


## Check BAMs

We can use functions from the package 'ATACseqQC' to inspect our BAM files and generate some ATAC-specific QC plots. For more information on how to use this package see https://haibol2016.github.io/ATACseqQCWorkshop/articles/ATACseqQC_workshop.html. 

First, lets check our BAMs using the function `bamQC`:

```{r check_BAMs_OMNI}
library(ATACseqQC)

bamQC(OMNI_bam_path, index = OMNI_bai_path, outPath = NULL)

```

```{r check_BAMs_STD}

bamQC(STD_bam_path, index = STD_bai_path, outPath = NULL)

```

### Q: The BAM files have been subsetted so we can work with them locally, from the above output can you tell how they have been cut down? 
A: We can see that, for both BAM files, there are only reads mapped to chromosome Y. This is because we subset the BAM files to only include this chromosome. 


# Shifting Reads

A 9 base pair (bp) DNA insertion is introduced when the transposase-induced nick is repaired during the ATAC process. For some types of analysis we may want to have single base pair resolution. This means we will need to correct for this insertion by shifting our read coordinates. Reads aligned to the +ve strand should be offset by +4bp and to the -ve strand should be shifted by -5bp. For our purposes this shifting will not be neccessary, but if you need this step you it can be done by using the ATACseqQC function `shiftGAlignmentsList` 


# Quality Control

We have already had a look at the quality of our ATAC-seq data in the multi_qc report. After reads have been preprocessed and aligned, there are some additional ATAC-specific quality control visualisations we can generate. 


## Fragment Size Distribution

Good-quality ATACseq data should display a characteristic pattern when you plot the distribution of the fragment sizes. This is because DNA is packaged into nucleosomes, which are about 147bp in length and cannot be accessed by the Tn5 enzyme. This means that most Tn5 fragments  come from nucleosome-free regions (NFR), and therefore be <100bp long. Fragments can also be created if Tn5 enzymes insert just before and after 1, 2 or 3 nucleosomes. This results in an enrichment of fragments that are ~200, ~400 and ~600bp long. 

First we can look at this distribution for the OMNI sample:

```{r frag_size_OMNI}

fragSize_OMNI <- fragSizeDist(OMNI_bam_path, "OMNI", index = OMNI_bai_path)

```
Then in our STD sample:

```{r frag_size_STD}

fragSize_STD <- fragSizeDist(STD_bam_path, "STD", index = STD_bai_path)

```

## Loading annotations

Annotations are needed so we can see where the genes are in the genome. They can help us identify if our fragments map to gene bodies, transcription start sites, introns or intergenic regions. 


```{r load_TSS_annotations}

BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene")
library(TxDb.Hsapiens.UCSC.hg19.knownGene)

genesLocations <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
tssLocations <- resize(genesLocations,fix="start",width = 1)

print(tssLocations)

```
### Q: What kind of data format is our tssLocations object in?
A: The tssLocations object is a GRanges object which includes information about genomic ranges i.e. the name of the range, the chromosome the range is on and the start and end coordinates.

### Challenge: As our BAM files only include reads from the Y chromosome, we can cut down our annotation file to also only include genes from this chromosome. There are a few ways you can do this, one example is shown here:

```{r subset_annotations}

tssLocations <- tssLocations[seqnames(tssLocations) %in% "chrY"]

print(tssLocations)

```

## Transcription Start Site Enrichment

Another sign of good-quality ATACseq data is that most reads should map to transcription start sites (TSSs). This is because TSSs are transcriptionally active and therefore 'open'. For more information: https://rockefelleruniversity.github.io/RU_ATACseq/exercises/answers/ATACseq_part2_answers.html. We can use the function `regionPlot` from the package 'soGGi' to examine this. 

```{r TSS_enrichment}
library(soGGi)
library(ggplot2)

OMNI_plot <- regionPlot(bamFile = OMNI_bam_path, testRanges = tssLocations, paired = TRUE)
STD_plot <- regionPlot(bamFile = STD_bam_path, testRanges = tssLocations, paired = TRUE)

plotRegion(OMNI_plot) + theme_bw()
plotRegion(STD_plot) + theme_bw()

```
At this point if we see that one of our samples is very different from the others and has a very poor fragment size histogram or TSS enrichment we may determine it is poor quality and remove it from downstream analysis. In this case our OMNI and STD outputs look similar and good quality so we would keep both of them. 
